### PR TITLE
Do not rely on nodal flags

### PR DESCRIPTION
**Pull request by EZoni**
**Goal**

Do not rely on nodal flags such as `Ex_nodal_flag`, check the actual `IndexType` (staggering) of individual `MultiFab`s instead. Linked to issue #797.

**Description**

The nodal flags have been moved from the main WarpX class (static public members) to local variables inside the function `WarpX::AllocLevelMFs` (see comment below by @MaxThevenet). The function `getFieldNodalFlagData` in Source/Python/WarpXWrappers.cpp has also been re-designed (see comment below by @dpgrote).

**Note**

Changes in Source/Diagnostics/SliceDiagnostic.cpp made in this PR should not be relevant (provided that they do not break the current code), as this file will be deleted when old diagnostics will be deleted (see comment below by @RevathiJambunathan).

### CODE REVIEW

**Comment by dpgrote** [comment]
Thanks for the PR! In WarpXWrappers.cpp, you can change this:
`int* warpx_getEx_nodal_flag() {return getFieldNodalFlagData(WarpX::Ex_nodal_flag);}`
to this:
`int* warpx_getEx_nodal_flag() {return getFieldNodalFlagData(WarpX::GetInstance().getEfield(0, 0).ixType().toIntVect());}`

A cleanup with be to change getFieldNodalFlagData to accept a MultiFab and have it do the `mf.ixType().toIntVect()`.

**Comment by RevathiJambunathan** [comment]
@EZoni The Source/Diagnostics/SliceDiagnostic.cpp file will be deleted -- since it is part of old diagnostics. So no need to worry about index types defined there.

**Comment by EZoni** [comment]
@dpgrote @RevathiJambunathan Thank you for the useful feedback!

**Comment by MaxThevenet** [comment]
Thanks for this PR @EZoni!
It seems that `Ex_nodal_flag` etc. only depend on `WarpX::do_nodal` and macros, unless I missed something. If so, could `Ex_nodal_flag` etc. be local variables in `WarpX::AllocLevelMFs`?

**Comment by EZoni** [comment]
@MaxThevenet 

The nodal flags are still used to initialize the PML fields. I tried to fix this in commit 1d58519, but I got segfault errors when replacing, for example, `WarpX::Ex_nodal_flag` with `WarpX::GetInstance().getEfield_fp(0,0).ixType().toIntVect()` in `Source/BoundaryConditions/PML.cpp`. 

I have not yet figured out why this is happening, as `InitPML` (hence the PML constructor) is called after the calls `mypc->AllocData()` and `mypc->InitData()` in `WarpX::InitFromScratch`. If we do not find a way to fix this, the nodal flags will probably need to remain global as long as they are needed in the PML constructor.

PS: the PR is now ready for review, I was waiting for the last Travis CI build to pass before changing its status.

File: Source/Python/WarpXWrappers.cpp [hunk 1]
```diff
@@ -57,18 +57,19 @@ namespace
...
-         * The nodel flag data should not be modifiable from Python. */
-        int *nodel_flag_data = (int*) malloc(AMREX_SPACEDIM * sizeof(int));
+        const amrex::IntVect nodal_flag( mf.ixType().toIntVect() );
+        int *nodal_flag_data = (int*) malloc(AMREX_SPACEDIM * sizeof(int));
 
         constexpr int NODE = amrex::IndexType::NODE;
 
         for (int i=0 ; i < AMREX_SPACEDIM ; i++) {
-            nodel_flag_data[i] = (nodal_flag[i] == NODE ? 1 : 0);
+            nodal_flag_data[i] = (nodal_flag[i] == NODE ? 1 : 0);
```
Comment (MaxThevenet) [inline Source/Python/WarpXWrappers.cpp:70]: ^^

File: Source/BoundaryConditions/PML.cpp [hunk 2]
```diff
@@ -436,12 +436,18 @@ PML::PML (const BoxArray& grid_ba, const DistributionMapping& /*grid_dm*/,
...
 #endif
 
-    pml_E_fp[0].reset(new MultiFab(amrex::convert(ba,WarpX::Ex_nodal_flag), dm, 3, nge));
-    pml_E_fp[1].reset(new MultiFab(amrex::convert(ba,WarpX::Ey_nodal_flag), dm, 3, nge));
-    pml_E_fp[2].reset(new MultiFab(amrex::convert(ba,WarpX::Ez_nodal_flag), dm, 3, nge));
-    pml_B_fp[0].reset(new MultiFab(amrex::convert(ba,WarpX::Bx_nodal_flag), dm, 2, ngb));
-    pml_B_fp[1].reset(new MultiFab(amrex::convert(ba,WarpX::By_nodal_flag), dm, 2, ngb));
-    pml_B_fp[2].reset(new MultiFab(amrex::convert(ba,WarpX::Bz_nodal_flag), dm, 2, ngb));
+    pml_E_fp[0].reset( new MultiFab( amrex::convert( ba,
+        WarpX::GetInstance().getEfield_fp(0,0).ixType().toIntVect() ), dm, 3, nge ) );
```
Comment (MaxThevenet) [inline Source/BoundaryConditions/PML.cpp:440]: If this is too cumbersome, we could consider implementing functions `WarpX::ixTypeE(int dir)` etc.
Oh, now that I think about it, I am not sure that would work...

**Pull request merged**